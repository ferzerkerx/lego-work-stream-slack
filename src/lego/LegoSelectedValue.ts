import { LegoSelectedValueEntry } from './LegoSelectedValueEntry';

export class LegoSelectedValue {
  id: string;
  entries: LegoSelectedValueEntry[] = [];
}
