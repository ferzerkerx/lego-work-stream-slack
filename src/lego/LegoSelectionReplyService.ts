import { SlackAttachment, SlackMessage, Storage } from 'botkit';
import { LegoSelectMessage } from './LegoSelectMessage';
import { LegoSelectionService } from './LegoSelectionService';
import { LegoSelectedValue } from './LegoSelectedValue';

export class LegoSelectionReplyService {
  static createReply(
    message,
    storage: Storage<LegoSelectMessage>
  ): Promise<SlackMessage> {
    let fullMessageId: string = this.getMessageId(message);

    // @ts-ignore
    return storage.lego_messages
      .get(fullMessageId)
      .then((storedLegoMessage: LegoSelectMessage) => {
        const legoMessage = LegoSelectionService.createLegoSelectMessage({
          legoMessage: storedLegoMessage,
          fullMessageId,
          user: message.user,
          action: message.actions[0],
          channel: message.channel,
        });

        // @ts-ignore
        return storage.lego_messages.save(legoMessage).then(() => {
          return this._createReply(message, legoMessage);
        });
      })
      .catch(e => this.defaultErrorHandling(e));
  }

  private static getMessageId(message): string {
    return `${message.channel}_${message.original_message.ts}`;
  }

  private static _createReply(
    message,
    legoMessage: LegoSelectMessage
  ): SlackMessage {
    const reply: SlackMessage = message.original_message;

    let attachmentsToSend: SlackAttachment[] = reply.attachments.filter(
      attachment => attachment.callback_id === 'lego_stats'
    );

    let actionMappings: any = {};
    attachmentsToSend[0].actions.forEach(actionDetail => {
      actionMappings[actionDetail.name] = actionDetail.text;
    });

    attachmentsToSend.push({
      text: `${this.formatSelectedValues(
        legoMessage.selectedValues,
        actionMappings
      )}`,
    });

    reply.attachments = attachmentsToSend;

    return reply;
  }

  static formatSelectedValues(
    legoSelectedValues: LegoSelectedValue[],
    actionMappings: any
  ): string {
    let messageStr: string = '';

    for (let legoSelectedValue of legoSelectedValues) {
      messageStr += `${actionMappings[legoSelectedValue.id] ||
        legoSelectedValue.id}: `;
      for (let entry of legoSelectedValue.entries) {
        if (entry.value) {
          messageStr += `<@${entry.user}>:${entry.value}, `;
        }
      }
      messageStr += `\n`;
    }
    return messageStr;
  }

  private static defaultErrorHandling(err: Error): void {
    console.error(err);
  }
}
