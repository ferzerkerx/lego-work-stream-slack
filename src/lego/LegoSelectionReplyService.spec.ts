import { LegoSelectionReplyService } from './LegoSelectionReplyService';
import { LegoSelectedValue } from './LegoSelectedValue';

describe('LegoSelectionReplyService', () => {
  test('formatMessage properly', () => {
    let selectedValues: LegoSelectedValue[] = [
      {
        id: 'someActionName',
        entries: [{ user: 'someuser', value: 8 }],
      },
      {
        id: 'someOtherActionName',
        entries: [{ user: 'someuser', value: 6 }],
      },
    ];

    let actionMappings: any = {
      someActionName: 'theActualName',
    };

    const legoSelectMessage: string = LegoSelectionReplyService.formatSelectedValues(
      selectedValues,
      actionMappings
    );
    expect(legoSelectMessage).toMatchSnapshot();
  });
});
