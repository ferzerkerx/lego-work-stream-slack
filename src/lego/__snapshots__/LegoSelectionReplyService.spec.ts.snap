// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegoSelectionReplyService formatMessage properly 1`] = `
"theActualName: <@someuser>:8, 
someOtherActionName: <@someuser>:6, 
"
`;
